### ISSUE TITLE
What additional standard metadata items should be included in EventFlow?

### ISSUE THREAD

**Issue by karolz-ms** [Azure/diagnostics-eventflow#48]
The primary reason for attaching metadata to events is to enable outputs to produce various semantically-rich telemetry types. In general, metadata has no fixed schema, but for the sake of standardization and to ease output implementation EventFlow has the notion of "standard metadata". The standard metadata defines property names and types associated with well-known metadata tags. Currently two kinds of standard metadata are defined: "metric" and "request": https://github.com/Azure/diagnostics-eventflow#standard-metadata-types

We are planning to add some more standard metadata types and would appreciate feedback on what additional standard metadata types should be added to EventFlow, either to its core, or as a well-known metadata for one of its outputs?

For example, Microsoft Application Insights has notions of "dependency call", "event" (representing a significant user interaction or significant configuration change), "exception" (unexpected error) and "page view". The dependency call seems like a useful and generic concept that could be added to the core, but the others I am less sure about.

**Comment by nblumhardt** [comment]
In addition to the ones you've noted, I think `Timestamp` etc. also come under this heading (though the way they're represented may be different) - from EventData.cs:

```csharp
        public DateTimeOffset Timestamp { get; set; }
        public string ProviderName { get; set; }
        public LogLevel Level { get; set; }
        public long Keywords { get; set; }
```

Some observations based on the fields we used in Serilog and how this has mapped to the various back-ends and integrations:

 * `Timestamp` and `Level` concide with first-class metadata properties in Serilog 👍
 * `Message` is core to the representation of most log data; it'd be hard to reliably implement logging-oriented outputs without it
 * `Exception` has proven useful for some back-ends and for custom rendering - it's often given special treatment on the collection-side
 * Some notion of an event type is pretty common, but the semantics vary substantially
  - Serilog treats `MessageTemplate` as a surrogate event type (events logged with the same template have the same type), effectively globally distinct
  - MEL has both `EventId` (string or int), distinct per logger, and message templates (not always present, but global when it is)
  - NLog's about to get message template support
  - ETW etc. (as you'd have observed) has a pretty strong event id concept, distinct by source
 * `Source` appears almost universally in logging libraries - Serilog has a well-known, though not first-class property called `SourceContext` for this

The event type one is really powerful, but also hard to handle semantically. Not sure if there's a clean way to get end-to-end fidelity there.

FWIW, there's an interesting food-for-thought list in: https://github.com/nlog/nlog/wiki/Layout-Renderers.

**Comment by karolz-ms** [comment]
Very interesting stuff, thanks for taking time to share... but it is a bit different topic from the one that I attempted to start  😉 

You are talking about _standard event properties_.  My question was about standard metadata, which is a concept unique to EventFlow (or at least I am not aware of anything similar in Serilog, NLog or MEL). This concept is there because EventFlow is not attempting to be a logging library, but rather a "glue" that binds together diagnostic data from various sources and send it to various destinations. The metadata add more semantic information to events and helps derive richer telemetry types from them, ultimately enabling diagnostic backends to analyze and present your telemetry in interesting ways. For example, if you describe something as "request", the diagnostic backend might be able to automatically calculate reliability rate for various types of requests and show you that in a graph.

That said I do have a couple comments about what you said

- Looks like `ProviderName` in EventFlow is what you call `Source`. It is pretty much the same concept.
- Event ID is hard to get right as you said, that is why it wasn't included as first-class property in EventFlow. It can be a moving target too. For example, .NET `EventSource` now support schema-less logging (i.e. you can say `EventSource.Write(EventName, anonymous_data_object)` without defining your EventSource class. In this usage pattern `EventId` is dynamically generated and depends on the order of invocations of the Write() method with different event names. So you are forced to depend on `EventName` instead to uniquely identify an event... 
- `Message` and `MessageTemplate`... let me present a contrary view (no offense intended!) and say that `Message` is pretty much redundant. From practical perspective it is probably required to have some human-readable description of the event, but the important information is in event properties (together with event identity, which is the combination of `Source` plus some form of event ID), The properties are what can be machine-processed reliably and efficiently. If I am processing 10s or 100s of thousands events/sec, this is my primary concern. EventFlow unfortunately does not help you with this problem, but I cringe when I think how many times a message like "Service operation completed successfully" is being sent over the net. That is why `Message` is treated rather unceremoniously as just another payload property. Feel free to tell me that I am very, very wrong on this. 😃

**Comment by nblumhardt** [comment]
Thanks for the reply and explanation!

I can see your point regarding messages, if metrics/aggregate telemetry are the primary goal. I guess there's still the question to consider, if you're ingesting log data, of what a particular event payload means.

For example, with logs as input you could have two streams of events:

```csharp
Log.Information("Loading took {Elapsed} s", elapsed);
Log.Information("Saving took {Elapsed} s", elapsed);
```

Granted, the useful telemetry is the `Elapsed` property value, but any kind of analysis is going to need either the message, message template or an event id to make sense of this data. (Serilog pipelines would tend to use `"Loading took {Elapsed} s"` and `"Saving took {Elapsed} s"`, or numeric hashes of those strings, as the "event types" here.)

I might be in danger of derailing this thread, since I'm not 100% across the target scenarios/design goals, so I might switch to spectator mode 😎 but I'm watching with interest! Thanks again for the response.

**Comment by karolz-ms** [comment]
Yes, that makes perfect sense to me. In Serilog case `Message`, or message template, serves a double duty of being a template for the human-readable event description, but also provides event identity. Because of that it is indispensable. 

I very much appreciate your thoughts, please do share, no need to be a spectator! And thank you.

**Comment by karolz-ms** [comment]
@cwe1ss, any thoughts on this topic?

**Comment by cwe1ss** [comment]
I'm not very familiar with the internals of this project so excuse me if this is not very helpful:

I'm surprised that there's no generic "event"-type right now. Are e.g. all log calls from Microsoft.Extensions.Logging translated into "request" calls right now?

I think "event" should be the most generic one or the fallback, all others could be seen as special kinds of events.

I also like "request" and "exception" from Application Insights. Requests for doing request rates etc and exceptions because they are usually handled differently (e.g. For alerts, reports, ...)

I'm not sure about "dependency call" and "page view" because they are quite specialized. It might be hard for Inputs to translate incoming events to these. is there even any other input than Application Insights that could send these?

In other words, I like "event", "metric", "request", "exception".

**Comment by cwe1ss** [comment]
Off topic regarding `Timestamp` from above: Has there been a discussion about DateTime vs DateTimeOffset? I [looked at this in more detail](https://github.com/opentracing/opentracing-csharp/issues/4) in one of my projects and it seems to me like there's no value in using DateTimeOffset in systems that work in UTC only.

**Comment by karolz-ms** [comment]
Hey Christian, thanks a bunch for quick reply!

There is absolutely a generic "event" type in EventFlow, in fact, everything is an "event", it is just that by adding some metadata to it the "event" can become something else, or be re-interpreted.

We already have "request". I hear you and agree on "exception" being a good candidate and "page view" being not-so-good one. 

I am a little surprised by your comment on "dependency call" though. I think of it as a almost mirror image of "request", that is, it represents the event of calling some dependency from the service code. Could be another of your services or something external. Typically you want to know what you have called, how long it took to receive a response and whether the call was successful. Highlighting these kinds of events is very useful when you want to determine whether it is a dependency that is causing problems in your service, or is it something internal. Also gives you opportunity to be proactive about dependency that starts having trouble. Seems a pretty generic concept to me.

Regarding `DateTimeOffset` I agree that in the cloud world pretty much every time value is expected to be UTC and so the use of `DateTimeOffset` is not really necessary. That said, it also eliminates any ambiguity and it does not have any serious drawbacks, so I am inclined to keep it as-is.

**Comment by cwe1ss** [comment]
> I am a little surprised by your comment on "dependency call" though. I think of it as a almost mirror image of "request", that is, it represents the event of calling some dependency from the service code. 

I agree that it is very useful - especially how it is used in Application Insights. My concern was just, how many inputs can actually make use of it and if it's enough to justify the type. If you take e.g. ILogger from Microsoft.Extensions.Logging, the only generic factors available are the level and whether or not you have an exception argument. This can probably only be translated to "event" and "exception".

However, actual requests or dependency calls are probably not tracked with ILogger anyway. Instead they probably use something like ETW, DiagnosticSource, etc. But even with those there's still no easy way for the EventFlow input to figure out whether an event is a request, a dependency call or something else - It's just a name like "System.Net.Http.Out" that defines the type/semantics. 

But you probably already have the same problem with "request", so adding "dependency call" as well might be a good thing - if only just to support the Application Insights input. 

And if output systems can't handle it, they can map it back to a generic event anyway.

So I change my mind to "event", "exception", "request", "dependencyCall" (or maybe just "dependency"?!) :)

**Comment by karolz-ms** [comment]
Even with ILogger-style API the metadata can be applied fairly precisely as long as you include some properties into your event. For example, if you log something like this

`logger.Information("{action}: a call to {url} completed in {time}, result was {isSuccess}", "Dependency", "http://foo.com/api/doStuff", 34.56, true);`

then in eventFlowConfig you can match these properties like this:

```json
{
    "type": "metadata",
    "metadata": "dependency",
    "include": "action == Dependency",
    "urlProperty": "url",
    "durationProperty": "time",
    "isSuccessProperty": "isSuccess"
}
```
The include condition is an expression https://github.com/Azure/diagnostics-eventflow#logical-expressions

This will literally work as implemented today, although no standard input will understand this "dependency" metadata (so it will be ignored and you will still get a regular "event" out of this, like you said above). 

And yes, we do have "metric" already.

Much appreciate your input, cheers!

**Comment by karolz-ms** [comment]
Resolved with b10d8f7b9550e8594e7c95933895a0593525c0b2

"dependency" and "exception" became standard metadata
"event" is ApplicationInsight-specific concept

Documentation update will follow shortly

**Comment by jfloodnet** [comment]
Hi - so was this suppose to resolve #46 ? I'm still seeing exceptions logged as traces.

**Comment by karolz-ms** [comment]
@jfloodnet could you share an example?

**Comment by jfloodnet** [comment]
Hi - looks like my issue is related to #90. Will ask in there.
